File: algorithms/17. Letter Combinations of a Phone Number/solution.py
class Solution(object):
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        m = [" ", "*", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"]
        res = collections.deque()
        res.append('')
        for d in digits:
            for i in xrange(len(res)):
                s = res.popleft()
                for c in m[ord(d) - ord('0')]:
                    res.append(s + c)
        return list(res) if digits else []

File: algorithms/17. Letter Combinations of a Phone Number/solution.cc
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        vector<string> m({" ", "*", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"});
        vector<string> res({""});
        for ( char d : digits ) {
            for ( int n = res.size(), i = 0; i < n; ++i ) {
                for ( char c : m[d - '0'] ) {
                    res.push_back(res[i] + c);
                }
                res[i] = res.back();
                res.pop_back();
            }
        }
        return digits.size() ? res : vector<string> ();
    }
};
